### PR TITLE
(#518) Feature: prebuilt binaries

### PR DESCRIPTION
**Pull request by s1hofmann**
This PR is concerning the `opencv4nodejs` build setup:

- It enables builds for all supported major node releases (6, 7, 8, 9, 10 and latest)
- It introduces an additional platform and now includes macOS
- It introduces a new CI task called `prebuild`, which runs whenever a new tag is pushed
  - It pre-builds bindings for every major node version / platform
  - Adds these bindings to the release tag on GitHub
  - Skips building bindings on install whenever there's a suitable pre-built one available for the given platform / node ABI

Using pre-built bindings leads to much faster installation:

`OPENCV4NODEJS_DISABLE_AUTOBUILD=1 npm i opencv4nodejs`
```
real	3m33.047s
user	3m9.170s
sys	0m18.730s
```

`OPENCV4NODEJS_DISABLE_AUTOBUILD=1 npm i opencv4nodejs-prebuilt`
```
real	0m3.678s
user	0m2.670s
sys	0m0.830s
```

## Requirements

This PR requires a GitHub token which needs to be available via `GITHUB_TOKEN` env var.

### CODE REVIEW

File: ci/prebuild/prebuild.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+if [[ "$TRAVIS_OS_NAME" == "linux" ]];
+then
+    unset OPENCV4NODEJS_DISABLE_AUTOBUILD
```
Comment (justadudewhohacks) [inline ci/prebuild/prebuild.sh]: Have you actually tried this out? I can imagine, building opencv from scratch for every single build task in the matrix could result in very long build times.

**Comment by justadudewhohacks** [comment]
Hi @s1hofmann ,

Thanks alot for this contribution! Do you know how prebuilt-install works with napi versions? Asking in case we want to migrate opencv4nodejs to napi at some point.

There is a single testcase failing on appveyor. I can't really figure out why that is right now. Could you disable this testcase for now via an environment variable for that build job?

File: ci/prebuild/prebuild.sh (hunk 1 above)
Comment (s1hofmann) [reply ci/prebuild/prebuild.sh]: In my fork I'm only running prebuild tasks and yes, the job takes quite some time.
Funny enough, building OpenCV from scratch for each Linux job is (on average) actually faster than installing OpenCV@3 via Homebrew for macOS builds.
Screenshot 2019-03-13 at 09 31 56

Comment (justadudewhohacks) [reply ci/prebuild/prebuild.sh]: Oh okay, ~20min is actually quite good. I think that should be fine.

**Comment by s1hofmann** [comment]
Hi @justadudewhohacks,

it's possible to build for multiple N-API versions.
I'll be working with such a setup in the near future, so if opencv4nodejs will switch to N-API, I'd happy to help again!
It would also be possible to build against Electron or NW.js headers, but I haven't looked into these things yet.

I'll update the Appveyor config to deal with the failing build.

**Comment by s1hofmann** [comment]
Today I spent some time looking into the segfaults on macOS.

```
PID 47533 received SIGSEGV for address: 0x0
0   segfault-handler.node               0x00000001026d5318 _ZL16segfault_handleriP9__siginfoPv + 312
1   libsystem_platform.dylib            0x00007fff7e958b3d _sigtramp + 29
2   ???                                 0x0000000103824988 0x0 + 4353837448
3   libopencv_objdetect.3.4.dylib       0x00000001048dd7a8 _ZNK2cv13HOGDescriptor6copyToERS0_ + 136
4   opencv4nodejs.node                  0x0000000105b326bd _ZN2cv13HOGDescriptorC2ERKS0_ + 333
5   opencv4nodejs.node                  0x0000000105b3255d _ZN2cv13HOGDescriptorC1ERKS0_ + 29
6   opencv4nodejs.node                  0x0000000105b32530 _ZN13HOGDescriptor15getNativeObjectEv + 32
7   opencv4nodejs.node                  0x0000000105b32503 _ZN21InstanceConverterTypeI13HOGDescriptorN2cv13HOGDescriptorEE6unwrapEN2v85LocalINS4_5ValueEEE + 99
8   opencv4nodejs.node                  0x0000000105b32494 _ZN19SingleTypeConverterI21InstanceConverterTypeI13HOGDescriptorN2cv13HOGDescriptorEES3_E6unwrapEN2v85LocalINS6_5ValueEEE + 36
9   opencv4nodejs.node                  0x0000000105b303c4 _ZN17AbstractConverterI19SingleTypeConverterI21InstanceConverterTypeI13HOGDescriptorN2cv13HOGDescriptorEES4_ES4_E6unwrapEN2v85LocalINS8_5ValueEEE + 36
10  opencv4nodejs.node                  0x0000000105b2b615 _ZN13HOGDescriptor7ComputeERKN3Nan20FunctionCallbackInfoIN2v85ValueEEE + 85
11  segfault-handler.node               0x00000001026d6ff3 _ZN3Nan3impL23FunctionCallbackWrapperERKN2v820FunctionCallbackInfoINS1_5ValueEEE + 1475
12  node                                0x000000010023663f _ZN2v88internal25FunctionCallbackArguments4CallEPNS0_15CallHandlerInfoE + 623
13  node                                0x0000000100235b81 _ZN2v88internal12_GLOBAL__N_119HandleApiCallHelperILb0EEENS0_11MaybeHandleINS0_6ObjectEEEPNS0_7IsolateENS0_6HandleINS0_10HeapObjectEEESA_NS8_INS0_20FunctionTemplateInfoEEENS8_IS4_EENS0_16BuiltinArgumentsE + 689
14  node                                0x0000000100235220 _ZN2v88internalL26Builtin_Impl_HandleApiCallENS0_16BuiltinArgumentsEPNS0_7IsolateE + 288
15  ???                                 0x00002ab99305be3d 0x0 + 46976523943485
[1]    47532 segmentation fault  BINDINGS_DEBUG=1 node node_modules/.bin/mocha --require globals.js --recursiv
```

I'm not (yet) an expert in debugging native node addons, but to me it looks like there's an access violation when invoking a `cv::HOGDescriptor`'s copy constructor.

**Comment by s1hofmann** [comment]
@justadudewhohacks According to #518 the failing tests seem to be a general issue on macOS. How should we proceed with this PR?

**Comment by justadudewhohacks** [comment]
The optimal solution would be, to figure out why the copy constructor is causing segmentation faults on MacOS. The easier solution would be, to simply disable the HOG descriptor test suite in the MacOS build.

**Comment by justadudewhohacks** [comment]
Apparently this issue is present on MacOSX and Linux, we didn't notice it on Linux yet, since it didn't occur so far in the builds using node version 6 for some reason.

However, the issue will be fixed with #542.

**Comment by s1hofmann** [comment]
Hi @justadudewhohacks, I merged upstream and did some cleanup, mind taking another look?

**Comment by justadudewhohacks** [comment]
Hi @s1hofmann,

Can you merge latest master in your branch again. I had to make some fixes yesterday and also changed the CI scripts. Sorry for inconvenience.

**Pull request closed without merging**